Objects/exceptions.c

```diff
@@ -118,21 +123,153 @@ static PyObject *
+        const Py_ssize_t len = PyTuple_Size(self->args);
+        seq = PyTuple_New(len);
+        if (seq == NULL) {
+            goto fail;
+        }
+        for (Py_ssize_t i=0; i < len; i++) {
+            repr = PyObject_Repr(PyTuple_GET_ITEM(self->args, i));
+            if (repr == NULL) {
+                goto fail;
+            }
+            PyTuple_SET_ITEM(seq, i, repr);
+        }
+        args = PyUnicode_Join(separator, seq);
+        Py_DECREF(seq);
+    }
+
+    if (PyMapping_Check(self->kwargs)) {
+        const Py_ssize_t len = PyMapping_Length(self->kwargs);
+        if (len == -1) {
+            goto fail;
+        }
+        seq = PyTuple_New(len);
+        items = PyMapping_Items(self->kwargs);
+        if (seq == NULL || items == NULL) {
+            goto fail;
+        }
+        it = PyObject_GetIter(items);
+        if (it == NULL) {
+            goto fail;
+        }
+        Py_ssize_t i = 0;
+        while ((item = PyIter_Next(it)) != NULL) {
+            if (!PyTuple_Check(item) || PyTuple_GET_SIZE(item) != 2) {
+                PyErr_SetString(PyExc_ValueError, "items must return 2-tuples");
+                goto fail;
+            }
+            key = PyTuple_GET_ITEM(item, 0);
+            value = PyTuple_GET_ITEM(item, 1);
+            repr = PyUnicode_FromFormat("%S=%R", key, value);
+            if (repr == NULL) {
+                goto fail;
+            }
+            PyTuple_SET_ITEM(seq, i, repr);
+            i++;
+            Py_DECREF(item);
+        }
+        kwargs = PyUnicode_Join(separator, seq);
+        Py_DECREF(seq);
+        Py_DECREF(items);
+        Py_DECREF(it);
+    }
+    Py_DECREF(separator);
+
+    if (args == NULL && kwargs == NULL) {
+        result = PyUnicode_FromFormat("%s()", name, kwargs);
+    } else if (kwargs == NULL || PyUnicode_GET_LENGTH(kwargs) == 0) {
+        result = PyUnicode_FromFormat("%s(%S)", name, args);
+    } else if (args == NULL || PyUnicode_GET_LENGTH(args) == 0) {
+        result = PyUnicode_FromFormat("%s(%S)", name, kwargs);
+    } else {
+        result = PyUnicode_FromFormat("%s(%S, %S)", name, args, kwargs);
+    }
+    Py_XDECREF(args);
+    Py_XDECREF(kwargs);
+    return result;
+
+fail:
+    Py_XDECREF(separator);
+    Py_XDECREF(args);
+    Py_XDECREF(kwargs);
+    Py_XDECREF(seq);
+    Py_XDECREF(repr);
+    Py_XDECREF(item);
+    Py_XDECREF(items);
+    Py_XDECREF(it);
+    Py_XDECREF(key);
+    Py_XDECREF(value);
+    return NULL;
 }
 
 /* Pickling support */
 static PyObject *
 BaseException_reduce(PyBaseExceptionObject *self, PyObject *Py_UNUSED(ignored))
 {
-    if (self->args && self->dict)
-        return PyTuple_Pack(3, Py_TYPE(self), self->args, self->dict);
-    else
-        return PyTuple_Pack(2, Py_TYPE(self), self->args);
+    PyObject *functools;
+    PyObject *partial;
+    PyObject *constructor;
+    PyObject *result;
+    PyObject **newargs;
+
+    _Py_IDENTIFIER(partial);
+    functools = PyImport_ImportModule("functools");
```

Comment:
I tried to do that, I removed `__reduce__` and added `__getnewargs_ex__` to the methods as:

```c
static PyObject *
BaseException_getnewargs_ex(PyBaseExceptionObject *self, PyObject *Py_UNUSED(ignored))
{
    PyObject *args = PyObject_GetAttrString((PyObject *) self, "args");
    PyObject *kwargs = PyObject_GetAttrString((PyObject *) self, "kwargs");

    if (args == NULL || kwargs == NULL) {
        return NULL;
    }

    return Py_BuildValue("(OO)", args, kwargs);
}
```

but it brocke pickling. Did I miss something?

Reply:
Oh, found my mistake, using `__getnewargs_ex__` broke pickling for protocols 0 and 1. Is this expected?

I don't think this happened when using a partial reference on the the constructor of the class.

Reply:
Maybe it's ok to broke pickling support for protocols 0 and 1 since it was broken for keyword args anyway?

Reply:
Defining `__reduce_ex__` would let you restore the old behaviour for those protocols, but I'm not sure `__getnewargs_ex__` will still be called if you do that (we're reaching the limits of my own `pickle` knowledge).

@pitrou Do you have any recommendations here? (Context: trying to get `BaseException` to pickle keyword args properly, wanting to use `__getnewargs_ex__` for more recent pickle protocols, but wondering how to handle the older protocols that don't use that)

Reply:
No, `__reduce__` overrides all higher-level hooks such as `__getstate__` / `__setstate__` and `__getnewargs__`. (the `_ex` variants being similar here)

Reply:
Since `__reduce_ex__` is the most flexible approach, I don't think you want to care about `__getnewargs__`.

Reply:
Thanks @pitrou.

Looking at https://github.com/python/cpython/blob/5c8f537669d3379fc50bb0a96accac756e43e281/Objects/typeobject.c#L4414, I think what a `BaseException.__reduce_ex__` implementation could do is:

1. Define `__getnewargs_ex__`
2. For protocol 2 and higher, call up to `object.__reduce_ex__` (which will then invoke `__getnewargs_ex__`)
3. For protocols 0 and 1, delegate to the existing `BaseException.__reduce__`

Does that sound plausible, or have I missed something that will keep that approach from working?

Reply:
How should I call `object.__reduce_ex__`?

It seems to me that calling the builtin super is not done anywhere in the source code but I don't find the right way to do it.

Do I need to call `object___reduce_ex__` directly?

Reply:
@ncoghlan Well, I'm not sure why you wouldn't implement the entire logic in `__reduce_ex__`, instead of also defining `__getnewargs_ex__`?

Or, rather, you could just define `__getnewargs_ex__` and stop caring about protocols 0 and 1 (which are extremely obsolete by now, so we want to maintain compatibility, but fixing bugs in them is not important).

Reply:
@pitrou I only suggested delegating to `__getnewargs_ex__` because I wasn't sure how to mimic that behaviour from inside a custom `__reduce_ex__` implementation.

But if `__reduce__` still gets called for protocols 0 and 1 even when `__getnewargs_ex__` is defined, then that's even better.

Reply:
Hi @pitrou @ncoghlan, thanks for you input. I pushed a new commit that implement `__getnewargs_ex__` but it seems that `__reduce_ex__` does not check it and call `__reduce__` no matter what the protocol is:

```
>>> BaseException().__reduce_ex__(0)
(<class 'BaseException'>, ())
>>> BaseException().__reduce_ex__(1)
(<class 'BaseException'>, ())
>>> BaseException().__reduce_ex__(2)
(<class 'BaseException'>, ())
>>> BaseException().__reduce_ex__(3)
(<class 'BaseException'>, ())
>>> BaseException().__reduce_ex__(4)
(<class 'BaseException'>, ())
>>> BaseException().__getnewargs_ex__()
((), {})
```

If I remove the `__reduce__`, then it breaks pickling for protocols 0 and 1:
```
>>> BaseException().__reduce_ex__(0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/remi/src/cpython/Lib/copyreg.py", line 66, in _reduce_ex
    raise TypeError(f"cannot pickle {cls.__name__!r} object")
TypeError: cannot pickle 'BaseException' object
>>> BaseException().__reduce_ex__(1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/remi/src/cpython/Lib/copyreg.py", line 66, in _reduce_ex
    raise TypeError(f"cannot pickle {cls.__name__!r} object")
TypeError: cannot pickle 'BaseException' object
>>> BaseException().__reduce_ex__(2)
(<function __newobj__ at 0x105c63040>, (<class 'BaseException'>,), None, None, None)
>>> BaseException().__reduce_ex__(3)
(<function __newobj__ at 0x105c63040>, (<class 'BaseException'>,), None, None, None)
>>> BaseException().__reduce_ex__(4)
(<function __newobj__ at 0x105c63040>, (<class 'BaseException'>,), None, None, None)
```

Do I need to define a custom `__reduce_ex__` as well?

Reply:
I dug further and it seems my issue comes from https://github.com/python/cpython/blob/master/Lib/copyreg.py#L66, I will look into the details tomorrow.